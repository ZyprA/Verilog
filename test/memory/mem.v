module mem(clk, load, addr, d, q);
    input clk, load;
    input [15:0] addr, d;
    output [15:0] q;
    reg [15:0] q;
    reg [15:0] mem [4095:0];

    always @ (posedge clk)
        begin
            if (load) mem[addr] <= d;
            q <= mem[addr];
        end
endmodule

module mem_sim;
    reg clk, load;
    reg [15:0] addr, d;
    wire [15:0] q;

    mem m(clk, load, addr, d, q);

    initial begin
        $dumpfile("mem.vcd");
        $dumpvars(0, mem_sim);
    end

    initial begin
        clk = 0;
        forever #5 clk = ~clk;
    end

    initial begin
        load = 1; addr = 1; d = 23;
        #100 load = 0; addr = 0;
        #100 addr = 1; d = 1;
        #100 load = 1; addr = 1; d = 2;
        #100 load = 0; addr = 1;
        #100 $finish;
    end

endmodule

