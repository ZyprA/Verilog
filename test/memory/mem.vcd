$date
	Sun Feb  2 21:00:45 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module mem_sim $end
$var wire 16 ! q [15:0] $end
$var reg 16 " addr [15:0] $end
$var reg 1 # clk $end
$var reg 16 $ d [15:0] $end
$var reg 1 % load $end
$scope module m $end
$var wire 16 & addr [15:0] $end
$var wire 1 # clk $end
$var wire 16 ' d [15:0] $end
$var wire 1 % load $end
$var reg 16 ( q [15:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
bx (
b10111 '
b1 &
1%
b10111 $
0#
b1 "
bx !
$end
#5
1#
#10
0#
#15
b10111 !
b10111 (
1#
#20
0#
#25
1#
#30
0#
#35
1#
#40
0#
#45
1#
#50
0#
#55
1#
#60
0#
#65
1#
#70
0#
#75
1#
#80
0#
#85
1#
#90
0#
#95
1#
#100
0#
b0 "
b0 &
0%
#105
bx !
bx (
1#
#110
0#
#115
1#
#120
0#
#125
1#
#130
0#
#135
1#
#140
0#
#145
1#
#150
0#
#155
1#
#160
0#
#165
1#
#170
0#
#175
1#
#180
0#
#185
1#
#190
0#
#195
1#
#200
0#
b1 $
b1 '
b1 "
b1 &
#205
b10111 !
b10111 (
1#
#210
0#
#215
1#
#220
0#
#225
1#
#230
0#
#235
1#
#240
0#
#245
1#
#250
0#
#255
1#
#260
0#
#265
1#
#270
0#
#275
1#
#280
0#
#285
1#
#290
0#
#295
1#
#300
0#
b10 $
b10 '
1%
#305
1#
#310
0#
#315
b10 !
b10 (
1#
#320
0#
#325
1#
#330
0#
#335
1#
#340
0#
#345
1#
#350
0#
#355
1#
#360
0#
#365
1#
#370
0#
#375
1#
#380
0#
#385
1#
#390
0#
#395
1#
#400
0#
0%
#405
1#
#410
0#
#415
1#
#420
0#
#425
1#
#430
0#
#435
1#
#440
0#
#445
1#
#450
0#
#455
1#
#460
0#
#465
1#
#470
0#
#475
1#
#480
0#
#485
1#
#490
0#
#495
1#
#500
0#
