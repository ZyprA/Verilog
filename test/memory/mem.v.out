#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x117f05c20 .scope module, "mem_sim" "mem_sim" 2 15;
 .timescale 0 0;
v0x117f16460_0 .var "addr", 15 0;
v0x117f16510_0 .var "clk", 0 0;
v0x117f165a0_0 .var "d", 15 0;
v0x117f16670_0 .var "load", 0 0;
v0x117f16720_0 .net "q", 15 0, v0x117f16330_0;  1 drivers
S_0x117f05d90 .scope module, "m" "mem" 2 20, 2 1 0, S_0x117f05c20;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "load";
    .port_info 2 /INPUT 16 "addr";
    .port_info 3 /INPUT 16 "d";
    .port_info 4 /OUTPUT 16 "q";
v0x117f05fd0_0 .net "addr", 15 0, v0x117f16460_0;  1 drivers
v0x117f16070_0 .net "clk", 0 0, v0x117f16510_0;  1 drivers
v0x117f16110_0 .net "d", 15 0, v0x117f165a0_0;  1 drivers
v0x117f161b0_0 .net "load", 0 0, v0x117f16670_0;  1 drivers
v0x117f16250 .array "mem", 0 4095, 15 0;
v0x117f16330_0 .var "q", 15 0;
E_0x117f05b10 .event posedge, v0x117f16070_0;
    .scope S_0x117f05d90;
T_0 ;
    %wait E_0x117f05b10;
    %load/vec4 v0x117f161b0_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %load/vec4 v0x117f16110_0;
    %ix/getv 3, v0x117f05fd0_0;
    %ix/load 4, 0, 0; Constant delay
    %assign/vec4/a/d v0x117f16250, 0, 4;
T_0.0 ;
    %ix/getv 4, v0x117f05fd0_0;
    %load/vec4a v0x117f16250, 4;
    %assign/vec4 v0x117f16330_0, 0;
    %jmp T_0;
    .thread T_0;
    .scope S_0x117f05c20;
T_1 ;
    %vpi_call 2 23 "$dumpfile", "mem.vcd" {0 0 0};
    %vpi_call 2 24 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x117f05c20 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x117f05c20;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x117f16510_0, 0, 1;
T_2.0 ;
    %delay 5, 0;
    %load/vec4 v0x117f16510_0;
    %inv;
    %store/vec4 v0x117f16510_0, 0, 1;
    %jmp T_2.0;
    %end;
    .thread T_2;
    .scope S_0x117f05c20;
T_3 ;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x117f16670_0, 0, 1;
    %pushi/vec4 1, 0, 16;
    %store/vec4 v0x117f16460_0, 0, 16;
    %pushi/vec4 23, 0, 16;
    %store/vec4 v0x117f165a0_0, 0, 16;
    %delay 100, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x117f16670_0, 0, 1;
    %pushi/vec4 0, 0, 16;
    %store/vec4 v0x117f16460_0, 0, 16;
    %delay 100, 0;
    %pushi/vec4 1, 0, 16;
    %store/vec4 v0x117f16460_0, 0, 16;
    %pushi/vec4 1, 0, 16;
    %store/vec4 v0x117f165a0_0, 0, 16;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x117f16670_0, 0, 1;
    %pushi/vec4 1, 0, 16;
    %store/vec4 v0x117f16460_0, 0, 16;
    %pushi/vec4 2, 0, 16;
    %store/vec4 v0x117f165a0_0, 0, 16;
    %delay 100, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x117f16670_0, 0, 1;
    %pushi/vec4 1, 0, 16;
    %store/vec4 v0x117f16460_0, 0, 16;
    %delay 100, 0;
    %vpi_call 2 38 "$finish" {0 0 0};
    %end;
    .thread T_3;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "mem.v";
